src/compiler/builder.ts

```diff
@@ -1098,6 +1102,33 @@ namespace ts {
             );
         }
 
+        function getWriteFileUpdatingSignatureCallback(writeFile: WriteFileCallback | undefined): WriteFileCallback {
+            return (fileName, data, writeByteOrderMark, onError, sourceFiles, sourceMapUrlPos) => {
+                if (sourceFiles?.length === 1 && isDeclarationFileName(fileName)) {
+                    const file = sourceFiles[0];
+                    const info = state.fileInfos.get(file.resolvedPath)!;
+                    const signature = state.currentAffectedFilesSignatures?.get(file.resolvedPath) || info.signature;
+                    if (signature === file.version) {
+                        const newSignature = (computeHash || generateDjb2Hash)(sourceMapUrlPos !== undefined ? data.substring(0, sourceMapUrlPos) : data);
+                        if (newSignature !== file.version) { // Update it
+                            if (host.storeFilesChangingSignatureDuringEmit) (state.filesChangingSignature ||= new Set()).add(file.resolvedPath);
+                            if (state.exportedModulesMap) BuilderState.updateExportedModules(file, file.exportedModulesFromDeclarationEmit, state.currentAffectedFilesExportedModulesMap ||= BuilderState.createManyToManyPathMap());
+                            if (state.affectedFiles && state.affectedFilesIndex! < state.affectedFiles.length) {

```

Comment:
What is this condition? What’s the difference between storing the new signature in `state.currentAffectedFilesSignatures` and storing it in `info.signature`?

Reply:
when we are emitting for affectedFiles things are maintained in `currentAffectedFilesSignatures ` which are commited only after handling of all the affected files is completed.. 
This is  done so that affected files are handled correctly if the processing for them is cancelled midway, we can revert back as if we didnt process the change at all.. 
As part of #48784 i am looking into simplifying that esp now that we have affected files pending emit etc..  (this signature cache dates back when we didnt have all these things to track)